vignettes/namespace.Rmd

```diff
@@ -19,53 +19,118 @@ This is a little frustrating at first, but soon becomes second-nature.
+While S3 methods are regular functions with a special naming scheme, their "export" works a bit differently.
+S3 methods are exported only in the sense that calling the generic with the appropriate class will call the method; a user can't directly access the method definition by typing its name.
+A more technically correctly term would be to say that the method is **registered** so that the generics can find it.
 
-### S3 methods for generics in suggested packages
+You must register, i.e. `@export`, every S3 method regardless of whether or not the generic is exported.
+roxygen2 will warn you if you have forgotten.
 
-`@exportS3Method` tag allows you to generate `S3method()` namespace directives.
-Its primary use is for "delayed" method registration, which allows you to define methods for generics found in suggested packages (R \>= 3.6).
-For example,
+```{r}
+#' @export
+bizarro.character <- function(x, ...) {
+  letters <- strsplit(x, "")
+  letters_rev <- lapply(letters, rev)
+  vapply(letters_rev, paste, collapse = "", FUN.VALUE = character(1))
+}
+```
+
+You have four options for documenting an S3 method:
+
+-   Don't document it; it's not required, and not needed for simple generics where the user won't care about the details.
+-   If the method is particularly complex or has many arguments that the generic does not, you can document it in its own file. In this case, just document it as if its a function.
+-   You can use `@rdname` to document it with other methds for the generic. This is a good option if it's your generic, and you're providing a bunch of methods for different classes.
+-   You can use `@rdname` to document it with other methods for the class. This is typically the least appealing option because the different generics will have different arguments, leading to a cluttered and potentially confusing page.
+
+```{r}
+#' Take an object to bizarro world
+#' 
+#' @description
+#' This is an S3 generic. This package provides methods for the 
+#' following classes:
+#' 
+#' * `character`: reverses the order of the letters in each element of 
+#'    the vector.
+#' 
+#' @param x A vector.
+#' @export
+bizarro <- function(x, ...) {
+  UseMethod("bizarro")
+}
+
+#' @export
+#' @rdname bizarro
+bizarro.character <- function(x, ...) {
+  letters <- strsplit(x, "")
+  letters_rev <- lapply(letters, rev)
+  vapply(letters_rev, paste, collapse = "", FUN.VALUE = character(1))
+}
+```
+
+Typically, you will write methods for generics that are either defined in the current package or a package that is a hard dependency[^2] of your package.
+Sometimes, however, you will want to write a method for a suggested dependency.
+In this case, `@export` will not work because it assumes the generic is included or imported in your `NAMESPACE`. Instead, use `@exportS3Method`. This will use "delayed" method registration, which means the method will only be registered when the suggested package is loaded.
```

Comment:
So do we need to update this in R Packages 🤔 ? Scroll to the end of this section:

https://r-pkgs.org/dependencies-in-practice.html#imports-and-exports-related-to-s3

Reply:
Yeah, we should. We'll probably also want to sunset that `s3_register()` method in our own packages once we can depend on 3.6.  I'll file some issues so we don't forget.